fix: Make tests correctly run also on Windows

cdprete opened the pull request:
## Description
The tests have been reworked, where necessary, so that they correctly run also on Windows.
This PR fixes https://github.com/eclipse/jkube/issues/1338

## Type of change
 - [x] Bug fix (non-breaking change which fixes an issue)
 - [ ] Feature (non-breaking change which adds functionality)
 - [ ] Breaking change (fix or feature that would cause existing functionality to change
 - [x] Chore (non-breaking change which doesn't affect codebase;
   test, version modification, documentation, etc.)

## Checklist
 - [x] I have read the [contributing guidelines](https://www.eclipse.org/jkube/contributing)
 - [x] I signed-off my commit with a user that has signed the [Eclipse Contributor Agreement](https://www.eclipse.org/legal/ECA.php)
 - [x] I Added [CHANGELOG](../CHANGELOG.md) entry
 - [ ] I have implemented unit tests to cover my changes -> no need
 - [ ] I have updated the [documentation](../kubernetes-maven-plugin/doc) accordingly -> no need
 - [x] No new bugs, code smells, etc. in [SonarCloud](https://sonarcloud.io/dashboard?id=jkubeio_jkube) report
 - [ ] I tested my code in Kubernetes -> I can't test this at the moment
 - [ ] I tested my code in OpenShift -> I can't test this

manusa commented:
<!-- Eclipse JKube CI bot report -->

cdprete commented:
Refactoring the code with the Eclipse plugin for IntelliJ somehow breaks the compilation completely, since some classes can't be found anymore although they're there:
```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.8.1:compile (default-compile) on project jkube-kit-spring-boot: Compilation failure: Compilation failure:
[ERROR] /C:/Users/Cosimo Damiano Prete/IdeaProjects/jkube/jkube-kit/jkube-kit-spring-boot/src/main/java/org/eclipse/jkube/springboot/generator/SpringBootGenerator.java:[63,40] package Configs does not exist
[ERROR] /C:/Users/Cosimo Damiano Prete/IdeaProjects/jkube/jkube-kit/jkube-kit-spring-boot/src/main/java/org/eclipse/jkube/springboot/generator/SpringBootGenerator.java:[62,4] cannot find symbol
[ERROR]   symbol:   class AllArgsConstructor
[ERROR]   location: class org.eclipse.jkube.springboot.generator.SpringBootGenerator
[ERROR] -> [Help 1]
```
I don't have an explanation for this yet.

manusa commented on gradle-plugin/it/src/test/java/org/eclipse/jkube/gradle/plugin/tests/DependencyResourcesIT.java:
```diff
@@ -63,4 +68,20 @@ public void k8sResource_whenRunWithReplicas_generatesK8sManifestsIncludingDepend
         .contains("validating");
   }
+
+  private static String[] addCleanArgument(String... arguments) {
+    Builder<String> builder = Stream.builder();
+    /*
+     * For some weird reason, depending on the order of execution of the tests, the first running test locks the entire
+     * build folder even after it has finished running and, as a result, the next test can't clean such folder and it
+     * fails.
+     * Considering that the tests operate on different resources, it should be safe to skip the clean step completely.
+     */
+    if (!SystemUtils.IS_OS_WINDOWS) {
+      builder.add("clean");
+    }
+    Arrays.stream(arguments).forEach(builder);
+
+    return builder.build().toArray(String[]::new);
+  }
```
We'll dig a little bit more on this.

cdprete replied:
I did some researches about this, and it seems like that the Gradle daemon stays up even when the Gradle execution completes (well, it's a daemon after all), and this may be the issue.
I tried to use the [--stop](https://docs.gradle.org/current/userguide/gradle_daemon.html#sec:stopping_an_existing_daemon) flag to stop it, but it's not recognized for some reason (maybe it's not supported in version 6.9 you're using) and I don't want to hack the user's home with the `gradle.properties` file.

manusa replied:
I hope this is not the issue because we shouldn't stop the daemon between executions (we need to take advantage of that to make the tests quick).
I'll try to check this part on my machine later this week.

cdprete replied:
Wouldn't then be better to start it from Maven in the `pre-integration-test` phase and stop it in the `post-integration-phase` then? The real question is then how to connect/use it from the tests...

manusa replied:
We should just not tamper with the gradle executable and rely on the wrapper to work accordingly on the user's machine.

Just like you wouldn't normally kill your gradle daemons after a `gradle build` (or whatever) execution.

We could always use the `--no-daemon` command-line flag, but again, this would affect test execution performance. The idea would be to understand what's actually going on and why is the directory->folder locked.

manusa commented on gradle-plugin/it/src/test/java/org/eclipse/jkube/gradle/plugin/tests/FileSecretIT.java:
```diff
@@ -47,11 +55,46 @@ public void ocResource_whenRun_generatesK8sSecret() throws IOException, ParseExc
         .contains("validating");
   }
+
+  /**
+   * On Windows, the separator used across lines is "\r\n" in contrast to *nix system where it's "\n".
+   * This function expects as input the file used for comparison in the tests and, if the test(s) is(are) running on
+   * Windows, then re-encodes the secret in them so that the Base64 string is the same between the expected file and the
+   * generated one.
+   * This is necessary because the comparison file contains the Base64 string calculated from a *nix machine.
+   *
+   * @param expectedFile the file used for comparison
+   * @return a file which has the same content as {@code expectedFile}, but with the secret re-encoded
+   */
+  private File adjustSecretEncoding(File expectedFile) throws IOException {
```
We should probably fix this via .gitattributes so we don't have to mess with file manipulation in tests or in utility classes.

cdprete replied:
I'll give it a try. It's a good idea.

cdprete replied:
I checked this again and `.gitattributes` won't help since the secrets are already encoded in Base64 with the *nix line ending.
How would you use `.gitattributes` to re-encode the secret present in the file?

manusa commented on gradle-plugin/kubernetes/src/test/java/org/eclipse/jkube/gradle/plugin/task/KubernetesConfigViewTaskTest.java:
```diff
@@ -44,13 +57,14 @@ public void setUp() throws IOException {
     configViewTask.runTask();
     // Then
-    verify(taskEnvironment.logger, times(1))
-        .lifecycle(matches("k8s: \n---\noffline: true\nbuildStrategy: \"s2i\""));
```
Wouldn't it be more simpler to replace with a valid regex?
Something like:
```
.lifecycle(matches("k8s: \r?\n---\r?\noffline: true\r?\nbuildStrategy: \"s2i\""));
```

cdprete replied:
If I remember correctly, I adjusted this in the code generating these logs as well, because one line was using the `String.format` with `%n` in the string to print, but the other lines were using an hardcoded `\n`.
I can, of course, use such regex here, but I think that the logs should be consistent upfront (so, in the code which generates them).
Moreover, the logger gets called with other messages as well, in fact I'm quite surprised this test passed even on the CI in the past.

manusa replied:
This regex should work regardless of the inconsistency of line breaks.
Of course, the inconsistency source should also be changed.

> Moreover, the logger gets called with other messages as well, in fact I'm quite surprised this test passed even on the CI in the past.

Yes, but the verify matcher checks that the method was called exactly once with the provided regex, despite any of the other recorded calls.

cdprete replied:
Yep, but the error I was getting is that it was expecting that string but it got a completely different one. 
Looking at the code, seems like you log some other information before that line and it was trying to match the passed string again those information.
I'll give it again a try and report back to you the error in case the regex doesn't help.

manusa commented on jkube-kit/build/api/src/main/java/org/eclipse/jkube/kit/build/api/auth/AuthConfig.java:
```diff
@@ -60,23 +62,6 @@ public AuthConfig(String username, String password, String email, String auth) {
     this(username, password, email, auth, null);
   }
 
-  public String toHeaderValue(KitLogger logger) {
```
This class was completely reformatted and is really hard to track the changes (at least in the GH interface).
Could you please revert the reformatting so that it's clearer for us to see what was changed here?

manusa commented on jkube-kit/build/api/src/main/java/org/eclipse/jkube/kit/build/api/auth/RegistryAuth.java:
```diff
@@ -32,69 +33,69 @@
 @EqualsAndHashCode
 public class RegistryAuth {
 
-    public static final RegistryAuth EMPTY_REGISTRY_AUTH =
```
This class was completely reformatted and is really hard to track the changes (at least in the GH interface).
Could you please revert the reformatting so that it's clearer for us to see what was changed here?

manusa commented on jkube-kit/build/api/src/main/java/org/eclipse/jkube/kit/build/api/auth/RegistryAuthConfig.java:
```diff
@@ -13,103 +13,102 @@
  */
 package org.eclipse.jkube.kit.build.api.auth;
 
-import lombok.Builder;
```
This class was completely reformatted and is really hard to track the changes (at least in the GH interface).
Could you please revert the reformatting so that it's clearer for us to see what was changed here?

manusa commented on jkube-kit/build/api/src/main/java/org/eclipse/jkube/kit/build/api/auth/handler/FromConfigRegistryAuthHandler.java:
```diff
@@ -24,31 +24,31 @@
  * @author roland
  */
 public class FromConfigRegistryAuthHandler implements RegistryAuthHandler {
-    private final RegistryAuthConfig registryAuthConfig;
```
This class was completely reformatted and is really hard to track the changes (at least in the GH interface).
Could you please revert the reformatting so that it's clearer for us to see what was changed here?

manusa commented on jkube-kit/build/api/src/main/java/org/eclipse/jkube/kit/build/api/auth/handler/OpenShiftRegistryAuthHandler.java:
```diff
@@ -22,64 +24,61 @@
 import org.eclipse.jkube.kit.build.api.auth.RegistryAuthHandler;
 import org.eclipse.jkube.kit.common.KitLogger;
 
-import static org.eclipse.jkube.kit.build.api.helper.KubernetesConfigAuthUtil.readKubeConfigAuth;
```
This class was completely reformatted and is really hard to track the changes (at least in the GH interface).
Could you please revert the reformatting so that it's clearer for us to see what was changed here?

manusa commented on jkube-kit/build/api/src/main/java/org/eclipse/jkube/kit/build/api/auth/handler/SystemPropertyRegistryAuthHandler.java:
```diff
@@ -22,44 +22,43 @@
  */
 public class SystemPropertyRegistryAuthHandler implements RegistryAuthHandler {
 
-    private final RegistryAuthConfig registryAuthConfig;
```
This class was completely reformatted and is really hard to track the changes (at least in the GH interface).
Could you please revert the reformatting so that it's clearer for us to see what was changed here?

manusa commented on jkube-kit/build/api/src/main/java/org/eclipse/jkube/kit/build/api/helper/BuildUtil.java:
```diff
@@ -13,22 +13,25 @@
 import org.eclipse.jkube.kit.common.JavaProject;
 import org.eclipse.jkube.kit.config.image.build.BuildConfiguration;
 
-import javax.annotation.Nonnull;
```
This class was completely reformatted and is really hard to track the changes (at least in the GH interface).
Could you please revert the reformatting so that it's clearer for us to see what was changed here?

manusa commented on jkube-kit/build/api/src/main/java/org/eclipse/jkube/kit/build/api/helper/DockerFileUtil.java:
```diff
@@ -42,282 +42,288 @@
  * @since 21/01/16
  */
 public class DockerFileUtil {
-    private static final String ARG_PATTERN_REGEX = "\\$(?:\\{(.*)\\}|(.*))";
```
This class was completely reformatted and is really hard to track the changes (at least in the GH interface).
Could you please revert the reformatting so that it's clearer for us to see what was changed here?

cdprete commented on gradle-plugin/it/src/test/java/org/eclipse/jkube/gradle/plugin/tests/KubernetesExtensionIT.java:
```diff
@@ -38,7 +41,38 @@ public void k8sConfigView_containsAFullyDeserializedConfiguration() throws IOExc
+        changePathsSeparatorIfNeeded(readFileToString(gradleRunner.resolveFile("expected", "expected-config.yml"), UTF_8)),
         false);
   }
+
+  /**
+   * On Windows, the paths extracted from the console output have the Windows path separator in them (\), but the file
+   * used for comparison has been written with the *nix path separator (/).
+   * This function expects as input the text of the file used for comparison and, if the test(s) is(are) running on
+   * Windows, then it changes the path separators in the Citrus assertions from / to \.
+   * Moreover, such paths will be escaped (in order to not break the Yaml parser), and - as such - they will become \\.
+   *
+   * @param text the text contained in the file used for comparison
+   * @return the same text, but with the correct path separators in the Citrus assertions
+   */
+  private static String changePathsSeparatorIfNeeded(String text) {
+    if (!SystemUtils.IS_OS_WINDOWS) {
+      return text;
+    }
+
+    final char matcherChar = '@';
+    final StringBuilder builder = new StringBuilder(text);
+    boolean isInMatcherDefinition = false;
+    for (int i = 0; i < builder.length(); ++i) {
+      final char currentChar = builder.charAt(i);
+      if (currentChar == matcherChar) {
+        isInMatcherDefinition = !isInMatcherDefinition;
+      } else if (isInMatcherDefinition && currentChar == '/') {
+        builder.deleteCharAt(i);
+        builder.insert(i, "\\\\");
+      }
+    }
+
+    return builder.toString();
+  }
```
I'll give it a try, although I'm scared since the Yaml parser could break :)

cdprete commented:
> Hi @cdprete I think you did an awesome job here, but so far it's very hard for me to review. Could you please revert any formatting changes that simply add noise to this PR.

Yep, I'll try to manually revert those re-arranged sources. Your complain is absolutely right (I hate that as well when I get such changes :D)

cdprete commented:
Hi @manusa. 
I've addressed everything I could from my side. What else can I do to make this PR land into master?

manusa commented on jkube-kit/common/src/test/java/org/eclipse/jkube/kit/common/util/EnvUtilTest.java:
```diff
@@ -387,7 +387,7 @@ public void testStoreTimestamp() throws IOException {
     public void testLoadTimestampShouldLoadFromFile() throws Exception {
         // Given
-        final File file = new File(EnvUtilTest.class.getResource("/util/loadTimestamp.timestamp").getFile());
+        final File file = new File(EnvUtilTest.class.getResource("/util/loadTimestamp.timestamp").toURI());
```
Why is this needed?

manusa replied:
```
C:\Users\mnurisan\00_BIN\jdk-17.0.1+12\bin\java.exe -ea -javaagent:C:\Users\mnurisan\.m2\repository/org/jmockit/jmockit/1.49/jmockit-1.49.jar -Didea.test.cyclic.buffer.size=1048576 -javaagent:C:\Users\mnurisan\00_BIN\ideaIU-2021.2.3\lib\idea_rt.jar=59982:C:\Users\mnurisan\00_BIN\ideaIU-2021.2.3\bin -Dfile.encoding=UTF-8 -classpath C:\Users\mnurisan\00_BIN\ideaIU-2021.2.3\lib\idea_rt.jar;C:\Users\mnurisan\00_BIN\ideaIU-2021.2.3\plugins\junit\lib\junit5-rt.jar;C:\Users\mnurisan\00_BIN\ideaIU-2021.2.3\plugins\junit\lib\junit-rt.jar;C:\Users\mnurisan\00-PROJECTS\forks\jkube\jkube-kit\common\target\test-classes;C:\Users\mnurisan\00-PROJECTS\forks\jkube\jkube-kit\common\target\classes;C:\Users\mnurisan\.m2\repository\io\fabric8\kubernetes-client\5.11.2\kubernetes-client-5.11.2.jar;C:\Users\mnurisan\.m2\repository\io\fabric8\kubernetes-model-core\5.11.2\kubernetes-model-core-5.11.2.jar;C:\Users\mnurisan\.m2\repository\io\fabric8\kubernetes-model-common\5.11.2\kubernetes-model-common-5.11.2.jar;C:\Users\mnurisan\.m2\repository\io\fabric8\kubernetes-model-rbac\5.11.2\kubernetes-model-rbac-5.11.2.jar;C:\Users\mnurisan\.m2\repository\io\fabric8\kubernetes-model-admissionregistration\5.11.2\kubernetes-model-admissionregistration-5.11.2.jar;C:\Users\mnurisan\.m2\repository\io\fabric8\kubernetes-model-apps\5.11.2\kubernetes-model-apps-5.11.2.jar;C:\Users\mnurisan\.m2\repository\io\fabric8\kubernetes-model-autoscaling\5.11.2\kubernetes-model-autoscaling-5.11.2.jar;C:\Users\mnurisan\.m2\repository\io\fabric8\kubernetes-model-apiextensions\5.11.2\kubernetes-model-apiextensions-5.11.2.jar;C:\Users\mnurisan\.m2\repository\io\fabric8\kubernetes-model-batch\5.11.2\kubernetes-model-batch-5.11.2.jar;C:\Users\mnurisan\.m2\repository\io\fabric8\kubernetes-model-certificates\5.11.2\kubernetes-model-certificates-5.11.2.jar;C:\Users\mnurisan\.m2\repository\io\fabric8\kubernetes-model-coordination\5.11.2\kubernetes-model-coordination-5.11.2.jar;C:\Users\mnurisan\.m2\repository\io\fabric8\kubernetes-model-discovery\5.11.2\kubernetes-model-discovery-5.11.2.jar;C:\Users\mnurisan\.m2\repository\io\fabric8\kubernetes-model-events\5.11.2\kubernetes-model-events-5.11.2.jar;C:\Users\mnurisan\.m2\repository\io\fabric8\kubernetes-model-extensions\5.11.2\kubernetes-model-extensions-5.11.2.jar;C:\Users\mnurisan\.m2\repository\io\fabric8\kubernetes-model-flowcontrol\5.11.2\kubernetes-model-flowcontrol-5.11.2.jar;C:\Users\mnurisan\.m2\repository\io\fabric8\kubernetes-model-networking\5.11.2\kubernetes-model-networking-5.11.2.jar;C:\Users\mnurisan\.m2\repository\io\fabric8\kubernetes-model-metrics\5.11.2\kubernetes-model-metrics-5.11.2.jar;C:\Users\mnurisan\.m2\repository\io\fabric8\kubernetes-model-policy\5.11.2\kubernetes-model-policy-5.11.2.jar;C:\Users\mnurisan\.m2\repository\io\fabric8\kubernetes-model-scheduling\5.11.2\kubernetes-model-scheduling-5.11.2.jar;C:\Users\mnurisan\.m2\repository\io\fabric8\kubernetes-model-storageclass\5.11.2\kubernetes-model-storageclass-5.11.2.jar;C:\Users\mnurisan\.m2\repository\io\fabric8\kubernetes-model-node\5.11.2\kubernetes-model-node-5.11.2.jar;C:\Users\mnurisan\.m2\repository\com\squareup\okhttp3\okhttp\3.12.12\okhttp-3.12.12.jar;C:\Users\mnurisan\.m2\repository\com\squareup\okio\okio\1.15.0\okio-1.15.0.jar;C:\Users\mnurisan\.m2\repository\com\squareup\okhttp3\logging-interceptor\3.12.12\logging-interceptor-3.12.12.jar;C:\Users\mnurisan\.m2\repository\io\fabric8\zjsonpatch\0.3.0\zjsonpatch-0.3.0.jar;C:\Users\mnurisan\.m2\repository\com\github\mifmif\generex\1.0.2\generex-1.0.2.jar;C:\Users\mnurisan\.m2\repository\dk\brics\automaton\automaton\1.11-8\automaton-1.11-8.jar;C:\Users\mnurisan\.m2\repository\io\fabric8\openshift-client\5.11.2\openshift-client-5.11.2.jar;C:\Users\mnurisan\.m2\repository\io\fabric8\openshift-model\5.11.2\openshift-model-5.11.2.jar;C:\Users\mnurisan\.m2\repository\io\fabric8\openshift-model-clusterautoscaling\5.11.2\openshift-model-clusterautoscaling-5.11.2.jar;C:\Users\mnurisan\.m2\repository\io\fabric8\openshift-model-operator\5.11.2\openshift-model-operator-5.11.2.jar;C:\Users\mnurisan\.m2\repository\io\fabric8\openshift-model-operatorhub\5.11.2\openshift-model-operatorhub-5.11.2.jar;C:\Users\mnurisan\.m2\repository\io\fabric8\openshift-model-machine\5.11.2\openshift-model-machine-5.11.2.jar;C:\Users\mnurisan\.m2\repository\io\fabric8\openshift-model-whereabouts\5.11.2\openshift-model-whereabouts-5.11.2.jar;C:\Users\mnurisan\.m2\repository\io\fabric8\openshift-model-monitoring\5.11.2\openshift-model-monitoring-5.11.2.jar;C:\Users\mnurisan\.m2\repository\io\fabric8\openshift-model-storageversionmigrator\5.11.2\openshift-model-storageversionmigrator-5.11.2.jar;C:\Users\mnurisan\.m2\repository\io\fabric8\openshift-model-tuned\5.11.2\openshift-model-tuned-5.11.2.jar;C:\Users\mnurisan\.m2\repository\io\fabric8\openshift-model-console\5.11.2\openshift-model-console-5.11.2.jar;C:\Users\mnurisan\.m2\repository\io\fabric8\openshift-model-machineconfig\5.11.2\openshift-model-machineconfig-5.11.2.jar;C:\Users\mnurisan\.m2\repository\io\fabric8\openshift-model-miscellaneous\5.11.2\openshift-model-miscellaneous-5.11.2.jar;C:\Users\mnurisan\.m2\repository\org\apache\commons\commons-lang3\3.11\commons-lang3-3.11.jar;C:\Users\mnurisan\.m2\repository\org\apache\commons\commons-compress\1.21\commons-compress-1.21.jar;C:\Users\mnurisan\.m2\repository\commons-io\commons-io\2.8.0\commons-io-2.8.0.jar;C:\Users\mnurisan\.m2\repository\com\google\code\gson\gson\2.8.6\gson-2.8.6.jar;C:\Users\mnurisan\.m2\repository\com\google\guava\guava\30.1.1-jre\guava-30.1.1-jre.jar;C:\Users\mnurisan\.m2\repository\com\google\guava\failureaccess\1.0.1\failureaccess-1.0.1.jar;C:\Users\mnurisan\.m2\repository\com\google\guava\listenablefuture\9999.0-empty-to-avoid-conflict-with-guava\listenablefuture-9999.0-empty-to-avoid-conflict-with-guava.jar;C:\Users\mnurisan\.m2\repository\com\google\code\findbugs\jsr305\3.0.2\jsr305-3.0.2.jar;C:\Users\mnurisan\.m2\repository\org\checkerframework\checker-qual\3.8.0\checker-qual-3.8.0.jar;C:\Users\mnurisan\.m2\repository\com\google\errorprone\error_prone_annotations\2.5.1\error_prone_annotations-2.5.1.jar;C:\Users\mnurisan\.m2\repository\com\google\j2objc\j2objc-annotations\1.3\j2objc-annotations-1.3.jar;C:\Users\mnurisan\.m2\repository\com\fasterxml\jackson\core\jackson-core\2.12.4\jackson-core-2.12.4.jar;C:\Users\mnurisan\.m2\repository\com\fasterxml\jackson\core\jackson-databind\2.12.4\jackson-databind-2.12.4.jar;C:\Users\mnurisan\.m2\repository\com\fasterxml\jackson\core\jackson-annotations\2.12.4\jackson-annotations-2.12.4.jar;C:\Users\mnurisan\.m2\repository\com\fasterxml\jackson\dataformat\jackson-dataformat-yaml\2.12.4\jackson-dataformat-yaml-2.12.4.jar;C:\Users\mnurisan\.m2\repository\org\yaml\snakeyaml\1.27\snakeyaml-1.27.jar;C:\Users\mnurisan\.m2\repository\com\fasterxml\jackson\datatype\jackson-datatype-jsr310\2.12.4\jackson-datatype-jsr310-2.12.4.jar;C:\Users\mnurisan\.m2\repository\org\slf4j\slf4j-api\1.7.25\slf4j-api-1.7.25.jar;C:\Users\mnurisan\.m2\repository\org\eclipse\jgit\org.eclipse.jgit\5.12.0.202106070339-r\org.eclipse.jgit-5.12.0.202106070339-r.jar;C:\Users\mnurisan\.m2\repository\com\googlecode\javaewah\JavaEWAH\1.1.7\JavaEWAH-1.1.7.jar;C:\Users\mnurisan\.m2\repository\javax\validation\validation-api\2.0.1.Final\validation-api-2.0.1.Final.jar;C:\Users\mnurisan\.m2\repository\org\fusesource\jansi\jansi\2.3.3\jansi-2.3.3.jar;C:\Users\mnurisan\.m2\repository\org\javassist\javassist\3.20.0-GA\javassist-3.20.0-GA.jar;C:\Users\mnurisan\.m2\repository\org\projectlombok\lombok\1.18.22\lombok-1.18.22.jar;C:\Users\mnurisan\00-PROJECTS\forks\jkube\jkube-kit\common-test\target\classes;C:\Users\mnurisan\.m2\repository\junit\junit\4.13.1\junit-4.13.1.jar;C:\Users\mnurisan\.m2\repository\org\hamcrest\hamcrest-core\1.3\hamcrest-core-1.3.jar;C:\Users\mnurisan\.m2\repository\org\jmockit\jmockit\1.49\jmockit-1.49.jar;C:\Users\mnurisan\.m2\repository\org\assertj\assertj-core\3.18.0\assertj-core-3.18.0.jar;C:\Users\mnurisan\.m2\repository\org\hamcrest\hamcrest-all\1.3\hamcrest-all-1.3.jar;C:\Users\mnurisan\.m2\repository\com\networknt\json-schema-validator\1.0.53\json-schema-validator-1.0.53.jar com.intellij.rt.junit.JUnitStarter -ideVersion5 -junit4 org.eclipse.jkube.kit.common.util.EnvUtilTest,testLoadTimestampShouldLoadFromFile

Process finished with exit code 0
```

cdprete replied:
In your path you don't  have spaces. I do :)
With spaces, `getFile` doesn't work properly

cdprete commented:
> I recall I fixed the `jkube-kit-common` module a while ago to work under windows. I just run the tests in the master branch, and they are still passing (on my machine). I wouldn't expect any changes in that module.
> 
> Same for `jkube-kit-common-maven`, I'd expect only the minimum changes to make the tests pass.

Maybe new changes broke them again? I don't know what to tell you.
If I changed something there is because I needed to. I didn't just change things because I like to, especially in my private time :)

manusa commented:
> Maybe new changes broke them again? I don't know what to tell you. If I changed something there is because I needed to. I didn't just change things because I like to, especially in my private time :)

I ran the tests in master yesterday with no failures. This `In your path you don't have spaces. I do :)` might explain it. I'll try again with an env more similar to yours. But the end goal would be to keep the set of changes as minimal as possible.

cdprete commented:
> > Maybe new changes broke them again? I don't know what to tell you. If I changed something there is because I needed to. I didn't just change things because I like to, especially in my private time :)
> 
> I ran the tests in master yesterday with no failures. This `In your path you don't have spaces. I do :)` might explain it. I'll try again with an env more similar to yours. But the end goal would be to keep the set of changes as minimal as possible.

Spaces where encoded as `%20` and such value was then literally used, leading to non-existing paths.

cdprete commented:
@manusa please let me know how you want to proceed with this :)

rohanKanojia commented:
@cdprete : Marc is on PTO and would be back on Friday so it might take some time to proceed forward.

In the meantime, Could you please resolve conflicts and address license check failure? `mvn -N license:format`

cdprete commented:
> @cdprete : Marc is on PTO and would be back on Friday so it might take some time to proceed forward.
> 
> In the meantime, Could you please resolve conflicts and address license check failure? `mvn -N license:format`

Done :)

cdprete commented:
After the rebase with `master`, the test `org.eclipse.jkube.kit.config.service.kubernetes.JibBuildServiceBuildIntegrationTest#build_withLayersAndArtifact_shouldPerformJibBuild` is now failing with 
```
org.eclipse.jkube.kit.config.service.JKubeServiceException: Error when building JIB image

	at org.eclipse.jkube.kit.config.service.kubernetes.JibBuildService.buildSingleImage(JibBuildService.java:106)
	at org.eclipse.jkube.kit.config.service.AbstractImageBuildService.processImage(AbstractImageBuildService.java:53)
	at org.eclipse.jkube.kit.config.service.AbstractImageBuildService.build(AbstractImageBuildService.java:34)
	at org.eclipse.jkube.kit.config.service.kubernetes.JibBuildServiceBuildIntegrationTest.build_withLayersAndArtifact_shouldPerformJibBuild(JibBuildServiceBuildIntegrationTest.java:136)
Caused by: java.lang.IllegalStateException: java.util.concurrent.ExecutionException: java.nio.file.NoSuchFileException: C:\Users\COSIMO~1\AppData\Local\Temp\jib-core-application-layers-cache\layers\88574cb1d05e6ed7129d41c0dac03bc94af067f715a50523793ba5b4d10dec87\56caeb713c5ea325d824d1c28f72375b5c545bc00657baeba0bccddeaa873939
	at org.eclipse.jkube.kit.service.jib.JibServiceUtil.buildContainer(JibServiceUtil.java:108)
	at org.eclipse.jkube.kit.config.service.kubernetes.JibBuildService.buildSingleImage(JibBuildService.java:100)
	... 3 more
Caused by: java.util.concurrent.ExecutionException: java.nio.file.NoSuchFileException: C:\Users\COSIMO~1\AppData\Local\Temp\jib-core-application-layers-cache\layers\88574cb1d05e6ed7129d41c0dac03bc94af067f715a50523793ba5b4d10dec87\56caeb713c5ea325d824d1c28f72375b5c545bc00657baeba0bccddeaa873939
	at com.google.common.util.concurrent.AbstractFuture.getDoneValue(AbstractFuture.java:566)
	at com.google.common.util.concurrent.AbstractFuture.get(AbstractFuture.java:547)
	at com.google.common.util.concurrent.FluentFuture$TrustedFuture.get(FluentFuture.java:88)
	at com.google.cloud.tools.jib.builder.steps.StepsRunner.run(StepsRunner.java:219)
	at com.google.cloud.tools.jib.api.Containerizer.run(Containerizer.java:390)
	at com.google.cloud.tools.jib.api.JibContainerBuilder.containerize(JibContainerBuilder.java:598)
	at org.eclipse.jkube.kit.service.jib.JibServiceUtil.buildContainer(JibServiceUtil.java:100)
	... 4 more
Caused by: java.nio.file.NoSuchFileException: C:\Users\COSIMO~1\AppData\Local\Temp\jib-core-application-layers-cache\layers\88574cb1d05e6ed7129d41c0dac03bc94af067f715a50523793ba5b4d10dec87\56caeb713c5ea325d824d1c28f72375b5c545bc00657baeba0bccddeaa873939
	at java.nio.file.Files.newByteChannel(Files.java:361)
	at java.nio.file.Files.newByteChannel(Files.java:407)
	at java.nio.file.spi.FileSystemProvider.newInputStream(FileSystemProvider.java:384)
	at java.nio.file.Files.newInputStream(Files.java:152)
	at com.google.cloud.tools.jib.blob.FileBlob.writeTo(FileBlob.java:38)
	at com.google.cloud.tools.jib.tar.TarStreamBuilder.writeAsTarArchiveTo(TarStreamBuilder.java:53)
	at com.google.cloud.tools.jib.image.ImageTarball.dockerWriteTo(ImageTarball.java:166)
	at com.google.cloud.tools.jib.image.ImageTarball.writeTo(ImageTarball.java:84)
	at com.google.cloud.tools.jib.builder.steps.WriteTarFileStep.call(WriteTarFileStep.java:66)
	at com.google.cloud.tools.jib.builder.steps.StepsRunner.lambda$writeTarFile$19(StepsRunner.java:635)
	at com.google.common.util.concurrent.TrustedListenableFutureTask$TrustedFutureInterruptibleTask.runInterruptibly(TrustedListenableFutureTask.java:125)
	at com.google.common.util.concurrent.InterruptibleTask.run(InterruptibleTask.java:69)
	at com.google.common.util.concurrent.TrustedListenableFutureTask.run(TrustedListenableFutureTask.java:78)
	at java.lang.Thread.run(Thread.java:750)
	Suppressed: java.io.IOException: This archive contains unclosed entries.
		at org.apache.commons.compress.archivers.tar.TarArchiveOutputStream.finish(TarArchiveOutputStream.java:291)
		at org.apache.commons.compress.archivers.tar.TarArchiveOutputStream.close(TarArchiveOutputStream.java:309)
		at com.google.cloud.tools.jib.tar.TarStreamBuilder.writeAsTarArchiveTo(TarStreamBuilder.java:56)
		at com.google.cloud.tools.jib.image.ImageTarball.dockerWriteTo(ImageTarball.java:166)
		at com.google.cloud.tools.jib.image.ImageTarball.writeTo(ImageTarball.java:84)
		at com.google.cloud.tools.jib.builder.steps.WriteTarFileStep.call(WriteTarFileStep.java:66)
		at com.google.cloud.tools.jib.builder.steps.StepsRunner.lambda$writeTarFile$19(StepsRunner.java:635)
		at com.google.common.util.concurrent.TrustedListenableFutureTask$TrustedFutureInterruptibleTask.runInterruptibly(TrustedListenableFutureTask.java:125)
		at com.google.common.util.concurrent.InterruptibleTask.run(InterruptibleTask.java:69)
		at com.google.common.util.concurrent.TrustedListenableFutureTask.run(TrustedListenableFutureTask.java:78)
		at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
		at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
		... 1 more
```
Have been done changes recently on the JIB build?
I need to investigate this...

sunix commented:
hey @cdprete, any update on this?

cdprete commented:
Hi.
I didn't had time to check this out.
I hope I can try to find some time next weekend.

Il Ven 10 Giu 2022, 16:15 Sun S. D. Tan ***@***.***> ha
scritto:

> hey @cdprete <https://github.com/cdprete>, any update on this?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/eclipse/jkube/pull/1416#issuecomment-1152409495>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACAZGWC5GZW6HJTDQIYHGTDVONEZXANCNFSM5SMRVPOA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

cdprete commented:
@sunix after the rebase with master, the compilation now fails harder than before:
```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-plugin-plugin:3.5.1:helpmojo (help-goal) on project openshift-maven-plugin: Execution help-goal of goal org.apache.maven.plugins:maven-plugin-plugin:3.5.1:helpmojo failed: The source must not be a directory. -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal org.apache.maven.plugins:maven-plugin-plugin:3.5.1:helpmojo (help-goal) on project openshift-maven-plugin: Execution help-goal of goal org.apache.maven.plugins:maven-plugin-plugin:3.5.1:helpmojo failed: The source must not be a directory.
    at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute (MojoExecutor.java:306)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:211)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:165)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:157)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:121)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:127)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:294)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:960)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:293)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:196)
    at sun.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:498)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
Caused by: org.apache.maven.plugin.PluginExecutionException: Execution help-goal of goal org.apache.maven.plugins:maven-plugin-plugin:3.5.1:helpmojo failed: The source must not be a directory.
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:148)
    at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute (MojoExecutor.java:301)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:211)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:165)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:157)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:121)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:127)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:294)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:960)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:293)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:196)
    at sun.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:498)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
Caused by: org.codehaus.plexus.archiver.ArchiverException: The source must not be a directory.
    at org.codehaus.plexus.archiver.AbstractUnArchiver.validate (AbstractUnArchiver.java:185)
    at org.codehaus.plexus.archiver.AbstractUnArchiver.extract (AbstractUnArchiver.java:118)
    at org.apache.maven.tools.plugin.extractor.annotations.JavaAnnotationsMojoDescriptorExtractor.discoverClassesFromSourcesJar (JavaAnnotationsMojoDescriptorExtractor.java:225)
    at org.apache.maven.tools.plugin.extractor.annotations.JavaAnnotationsMojoDescriptorExtractor.scanJavadoc (JavaAnnotationsMojoDescriptorExtractor.java:177)
    at org.apache.maven.tools.plugin.extractor.annotations.JavaAnnotationsMojoDescriptorExtractor.execute (JavaAnnotationsMojoDescriptorExtractor.java:105)
    at org.apache.maven.tools.plugin.scanner.DefaultMojoScanner.populatePluginDescriptor (DefaultMojoScanner.java:96)
    at org.apache.maven.plugin.plugin.AbstractGeneratorMojo.execute (AbstractGeneratorMojo.java:259)
    at org.apache.maven.plugin.plugin.HelpGeneratorMojo.execute (HelpGeneratorMojo.java:92)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
    at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute (MojoExecutor.java:301)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:211)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:165)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:157)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:121)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:127)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:294)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:960)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:293)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:196)
    at sun.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:498)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
[ERROR]
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/PluginExecutionException
[ERROR]
[ERROR] After correcting the problems, you can resume the build with the command
[ERROR]   mvn <args> -rf :openshift-maven-plugin
```
Before I was randomly having that error. Now it's persistent...

cdprete commented:
@sunix all green. You can proceed if you like :)
To work around the error I wrote about before, I just triggered the build again as suggested in the output from Maven.
Weird enough, the k8s plugin doesn't exhibit this behavior even if they're the same.
Finally, the "source" they're referring to points to the `classes` folder under `target` (just some short debugging).

sunix commented on gradle-plugin/it/src/test/java/org/eclipse/jkube/gradle/plugin/tests/FileSecretIT.java:
```diff
@@ -46,10 +54,45 @@ public void ocResource_whenRun_generatesK8sSecret() throws IOException, ParseExc
+   * @param expectedFile the file used for comparison
+   * @return a file which has the same content as {@code expectedFile}, but with the secret re-encoded
+   */
+  private File adjustSecretEncoding(File expectedFile) throws IOException {
```
Minor maybe rename it `adjustSecretEncodingOnWindows`

sunix commented:
> I usually like to encapsulate and hide such edge cases, also because in the future the condition(s) triggering a case may change and, in such a case, you would need to rename (at least) the method to keep consistency. But I can rename it, of course. Not a big deal :) 

Yes not a big deal :)
To me it really depends: here it important to tell the reader "this is for Windows!".

cdprete commented on jkube-kit/build/api/src/test/java/org/eclipse/jkube/kit/build/api/assembly/AssemblyManagerTest.java:
```diff
@@ -13,208 +13,213 @@
-      logger.warn(anyString, (Object []) any); times = 0;
-    }};
-    //@formatter:on
+  private static File resourceTestFile(String path) throws URISyntaxException {
```
I can have a better look at this. 

The issue I've seen it's that we need the same things in multiple sub-modules.
I usually do a `test-utils` sub-module to collect all such things when I need to reuse them, but I don't know how you guys want to proceed. Maybe adding a new sub-module is not really an option.

cdprete replied:
@sunix please let me know how to proceed here. 
It's indeed duplicated but in different modules. `TestUtils` was added because I needed the same logic in 2 different test classes of the sub-module where it's in.

I would suggest that we merge this as it is and another ticket gets open for refactoring the tests in general.
For example, I've just pulled the `master` branch and the latest commit says something about the tests being converted from JUnit 4 to 5. Unfortunately, only one test was changed :-D

sunix commented:
@manusa could you have a look? (once you are back from PTO ;) )

sunix commented:
@cdprete hey, so we have discussed with the team and it is taking time for us (among the others tasks we are working on)
The problem is that it is a quite big PR. So either 
- you wait until we find some time to make the review
- or you split this PR into smaller PRs that we could review and merge one after the other.

I hope you understand. Let us know what's work for you and thanks again for contributing!

cdprete commented:
Hi.
I don't think it makes so much sense to split it into multiple PRs since,
in the end, it's just a small refactoring.
Moreover, it would take me quite some time if I should split the things
now. It would have been different if I was asked to do this from the
beginning.

Therefore, from my side you can put this on hold. I don't know how much
time I'll have to keep things up to date, though.

Il Lun 4 Lug 2022, 10:25 Sun S. D. Tan ***@***.***> ha
scritto:

> @cdprete <https://github.com/cdprete> hey, so we have discussed with the
> team and it is taking time for us (among the others tasks we are working on)
> The problem is that it is a quite big PR. So either
>
>    - you wait until we find some time to make the review
>    - or you split this PR into smaller PRs that we could review and merge
>    one after the other.
>
> I hope you understand. Let us know what's work for you and thanks again
> for contributing!
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/eclipse/jkube/pull/1416#issuecomment-1173511924>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACAZGWEOLOARNWZQXLGJLSTVSKNXPANCNFSM5SMRVPOA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

rohanKanojia commented:
@cdprete : Sorry for the late response on this one.

We have broken the issue into module-wise smaller refactors. Your PR helped me in creating the issues.

cdprete commented:
> @cdprete : Sorry for the late response on this one.
> 
> We have broken the issue into module-wise smaller refactors. Your PR helped me in creating the issues.

I'm glad it helped somehow and sad at the same time to see (by looking at the open issues in GitHub) that many of these tests are still broken after 2+ years.

rohanKanojia commented:
@cdprete : Sorry that we couldn't provide timely feedback. I understand that we still haven't fixed all Windows tests, but we will try again to fix them in small incremental steps.
